### ISSUE TITLE
Messages counters display incorrect number of messages; alerts not consistently dismissed

### ISSUE THREAD

**Issue by jwzimmer** [LLK/scratch-www#1497]
**TL;DR:** There are a couple things going on here contributing to the effect of the counters being unreliable & the alert dismissal being unreliable: (1) Dismissing an alert doesn't always permanently dismiss it; (2) The Messages value should never be less than 0, but somehow it currently can be; (3) The counter in the navbar is wrong if either counter from the body of the Messages page is wrong.

Chrome 60, Mac OS El Capitan, Staging

Dismissing an alert always reduces the value of the "Messages from the Scratch Team" counter while you are on the Messages page, but once you navigate away and back again, the counter may not stay decremented.

**Repro**
If you click an X to dismiss an alert, then go to the Explore page, then go back to the Messages page, the alert you just dismissed will have come back.

The "Messages" counter on the body of the Messages page should never have a value less than 0, but currently it's displaying a value of -2. (I'm not sure exactly what causes this.)

The navbar message counter seems to be set based on the "Messages" counter + "Messages from the Scratch Team" counter, so if one of those values is incorrect, the value in the navbar message counter is also incorrect.

<img width="986" alt="screen shot 2017-09-01 at 1 10 37 pm" src="https://user-images.githubusercontent.com/1320456/29980270-0cdb4084-8f17-11e7-8fc3-ca5894e84b3d.png">
<img width="979" alt="screen shot 2017-09-01 at 1 10 31 pm" src="https://user-images.githubusercontent.com/1320456/29980271-0cdeba66-8f17-11e7-83ef-80ff0e54f015.png">
<img width="961" alt="screen shot 2017-09-01 at 1 01 19 pm" src="https://user-images.githubusercontent.com/1320456/29980020-ce982e5a-8f15-11e7-9094-61228ffe029c.png">
<img width="994" alt="screen shot 2017-09-01 at 1 01 12 pm" src="https://user-images.githubusercontent.com/1320456/29980022-ce9dddf0-8f15-11e7-8b72-9e64a99143f2.png">

![out3](https://user-images.githubusercontent.com/1320456/29980516-1492ed9e-8f18-11e7-9685-f1810d8ff24d.gif)

**Comment by mewtaylor** [comment]
@jwzimmer this is due to the behavior of navigating away from the messages page, and immediately back to it within the span of 2 minutes – these api endpoints are all cached for 2 minutes. If you were to come back after 2 minutes, it would display correctly. These messages are indeed dismissed, but the list of messages will not get updated until after the cache expires. 

I don't think it's a good idea to purge these endpoints after a dismissal and optimize for immediate revisits, because this is a behavior that is more often seen in testing than in production. Consequently, I think we should not worry about the message re-appearing in this type of interaction.

I can look into the negative counter though, that seems like an actual bug.

**Comment by mewtaylor** [comment]
tracking in https://github.com/LLK/scratch-api/pull/325

**Comment by mewtaylor** [comment]
Fixed in https://github.com/LLK/scratch-api/pull/325

**Comment by chrisgarrity** [comment]
The first time I went to the messages page the number was wrong (7 not 3):
![screen shot 2017-09-05 at 4 34 16 pm](https://user-images.githubusercontent.com/399209/30082444-d53491c2-9258-11e7-8718-9c3e1b31a17c.png)

But the next time it was correct:
![screen shot 2017-09-05 at 4 38 58 pm](https://user-images.githubusercontent.com/399209/30082461-e5589da0-9258-11e7-9da4-d3e968ad7772.png)

By the way it is a student account.

**Comment by jwzimmer** [comment]
Currently I am seeing this with the account `abcdekfy` on Staging (Safari, Mac OS El Capitan).

<img width="1043" alt="screen shot 2017-09-06 at 2 00 23 pm" src="https://user-images.githubusercontent.com/1320456/30127083-fd0e3a96-930b-11e7-89a2-b9582525233b.png">
<img width="987" alt="screen shot 2017-09-06 at 2 00 06 pm" src="https://user-images.githubusercontent.com/1320456/30127084-fd11ed12-930b-11e7-9584-55f88c07e17b.png">

**Comment by mewtaylor** [comment]
@jwzimmer that most recent version of this issue should be solved in https://github.com/LLK/scratchr2/commit/e3da4b1a589271a05a723a093555b8e4c3b2a9de. As for the student account issue that was being seen by @chrisgarrity, I still have not repro'd yet

**Comment by jwzimmer** [comment]
The invite to become a Scratcher doesn't seem to be incrementing the messages counter in the navbar:

<img width="1075" alt="screen shot 2017-09-07 at 2 56 38 pm" src="https://user-images.githubusercontent.com/1320456/30180138-debacb96-93dc-11e7-933b-abf6d37d6149.png">

**Comment by jwzimmer** [comment]
Scratchr2 navbar said 0, but scratch-www navbar says 2 (for arglebargle):
<img width="999" alt="screen shot 2017-09-07 at 4 40 55 pm" src="https://user-images.githubusercontent.com/1320456/30184254-7c6a5600-93eb-11e7-93e9-d963e28b1a58.png">
<img width="1010" alt="screen shot 2017-09-07 at 4 40 50 pm" src="https://user-images.githubusercontent.com/1320456/30184256-7c6cbe54-93eb-11e7-9ef0-638568170372.png">
<img width="1032" alt="screen shot 2017-09-07 at 4 40 33 pm" src="https://user-images.githubusercontent.com/1320456/30184255-7c6a54c0-93eb-11e7-9cdd-2cd1a766c4a8.png">

**Comment by jwzimmer** [comment]
**Sort-of-related issue I noticed while trying to repro**

I saw 5 notifications marked as unread and 5 in the navbar counter, but the last notification was from 20 hours ago and had definitely been read multiple times today. 
@mewtaylor explained that:
> the way that works is that something appearing as “unread” is not a demarcation on the notification itself – there’s a count that’s cached, and it just shows the first `n` messages as “unread” depending on the count
> so, it’s not that that message is “unread” – it’s that the count seems to have gotten off somehow

After investigation it turned out that the issue was what happened when a project was added to the studio.
@mewtaylor said:
> I think I see the issue. I had a place where I increment the cache count that I forgot to remove yesterday when I was making changes
so it incremented the cache count without there being something to display
because I originally thought adding a project to a studio resulted in a notification

But all the studio activities are consolidated into one notification per day.

@mewtaylor is fixing this issue now! : )

<img width="1067" alt="screen shot 2017-09-08 at 4 25 46 pm" src="https://user-images.githubusercontent.com/1320456/30230590-f1eda43a-94b4-11e7-891e-bdd993be57af.png">

**Comment by jwzimmer** [comment]
Should be fixed by @mewtaylor's latest changes.
